### PR TITLE
Add routing API

### PR DESCRIPTION
**Pull request by Alch-Emi**
A lot of clients will probably end up having many different endpoints that clients can connect to.  It can get pretty tiring to do all the route parsing on your own, so this PR adds a nice API to do it for you.  This replaces the old single handler method, but you can still have a single handler for all your requests if you set a handler for the "/" route.

What do you think?

---
This change is [Reviewable](https://reviewable.io/reviews/panicbit/northstar/29)

### CODE REVIEW

**Comment by panicbit** [comment]
This is probably a good point to start discussing the vision, goals and non-goals for Northstar. Originally my idea was to be similar to hyper with regards to API design, i.e. fairly simple and barebones. In that world, more higher-level frameworks could build upon Northstar without having to deal with many Gemini specific details.

What is your idea of what Northstar currently is and should become in the future?

**Comment by Alch-Emi** [comment]
That is entirely valid!  I was definitely going for more of an ease-of-use focused approach, with the goal of taking as much work off of the user as possible, but I absolutely see the need for having a low-level library as well, and I wouldn't want to commandeer your vision.

I might actually split a few of the things I've been working on into their own library built on top of northstar, like you mentioned, 'cause I've been playing with a few features that I want to use in the project I've been (slowly) putting together.  You can kinda see that taking shape if you look at the [tangled git network](https://github.com/panicbit/northstar/network) I've got going on in my fork.

Most of the stuff I've been working on is pretty high level, and I was well aware that there was a good chance it was going to be out of scope even as I was writing it.  That's not to say you can't use that code if you want to, just that I knew what I was getting into.

tldr; if you want to keep northstar small i totally get it and will find other ways to work with some of the features i had planned, like building on top of northstar, like you suggested.  whatever you decide, it's totally cool <3

**Comment by Alch-Emi** [comment]
My main concern with building on top of Northstar at the minute is that libraries would have to either completely wrap most of Northstar's API, or ask their end users to use the northstar API, but use somewhat jenky tricks in order to extend it, which might limit the scope of the library.

For example, if a library wanted to extend northstar with ratelimiting, this would require two things:  1) keeping track of a small amount of data attached to the Server struct, and 2) being able to access requests before they reach the end user's code.  This could be accomplished by adding a sealed trait which extends the Northstar builder with something like a `serve_with_ratelimits(RatelimitConfig, Handler)` method, or by adding a `wrap_handler(RatelimitConfig, Handler) -> Handler`, the first of which is ergonomic for the user, but a little hard to combine with other extensions, and the latter is easy to combine, but a little annoying for the end user.  They could also completely wrap northstar's API, which would give a lot of control, but would also mean a lot of re-exposing northstar's methods, and also a bit of bulk.

One possible solution to this is to try to make northstar a little more modular, which is part of what I was hoping to do with the routes api, but it would also need something like a middleware API, at which point northstar is starting to look a bit like a high-level library itself, so it's not like that's not an approach without downsides either.

In the end, all of the approaches have downside, but they all have their upsides as well, so whatever you choose I'm down with

**Comment by panicbit** [comment]
I think northstar should (eventually) provide a "complete package" and here is why: There is not a lot that you can do with gemini. Common usecases are probably very well catered by a rich high-level API. Someone with an extreme use-case can probably fairly easily implement a gemini server that is tailor made for their use-case from scratch. Northstar could probably still provide some low-level building blocks (lower than the current public API) to help with such cases.

With that decision out of the way, I can finally merge this PR :)

**Comment by panicbit** [comment]
I have also now published version `0.4.0` on crates.io
